x/group/keeper/msg_server.go

```diff
@@ -0,0 +1,810 @@
+		return k.groupAccountTable.Update(ctx.KVStore(k.key), groupAccount)
+	}
+
+	err := k.doUpdateGroupAccount(ctx, req.Address, req.Admin, action, "group account decision policy updated")
+	if err != nil {
+		return nil, err
+	}
+
+	return &group.MsgUpdateGroupAccountDecisionPolicyResponse{}, nil
+}
+
+func (k Keeper) UpdateGroupAccountMetadata(goCtx context.Context, req *group.MsgUpdateGroupAccountMetadata) (*group.MsgUpdateGroupAccountMetadataResponse, error) {
+	ctx := types.UnwrapSDKContext(goCtx)
+	metadata := req.GetMetadata()
+
+	action := func(groupAccount *group.GroupAccountInfo) error {
+		groupAccount.Metadata = metadata
+		groupAccount.Version++
+		return k.groupAccountTable.Update(ctx.KVStore(k.key), groupAccount)
+	}
+
+	if err := assertMetadataLength(metadata, "group account metadata"); err != nil {
+		return nil, err
+	}
+
+	err := k.doUpdateGroupAccount(ctx, req.Address, req.Admin, action, "group account metadata updated")
+	if err != nil {
+		return nil, err
+	}
+
+	return &group.MsgUpdateGroupAccountMetadataResponse{}, nil
+}
+
+func (k Keeper) CreateProposal(goCtx context.Context, req *group.MsgCreateProposal) (*group.MsgCreateProposalResponse, error) {
+	ctx := types.UnwrapSDKContext(goCtx)
+	accountAddress, err := sdk.AccAddressFromBech32(req.Address)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "request group account")
+	}
+	metadata := req.Metadata
+	proposers := req.Proposers
+	msgs := req.GetMsgs()
+
+	if err := assertMetadataLength(metadata, "metadata"); err != nil {
+		return nil, err
+	}
+
+	account, err := k.getGroupAccountInfo(ctx.Context(), accountAddress.Bytes())
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "load group account")
+	}
+
+	g, err := k.getGroupInfo(ctx.Context(), account.GroupId)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "get group by account")
+	}
+
+	// Only members of the group can submit a new proposal.
+	for i := range proposers {
+		if !k.groupMemberTable.Has(ctx.KVStore(k.key), orm.PrimaryKey(&group.GroupMember{GroupId: g.GroupId, Member: &group.Member{Address: proposers[i]}})) {
+			return nil, sdkerrors.Wrapf(errors.ErrUnauthorized, "not in group: %s", proposers[i])
+		}
+	}
+
+	// Check that if the messages require signers, they are all equal to the given group account.
+	if err := ensureMsgAuthZ(msgs, accountAddress); err != nil {
+		return nil, err
+	}
+
+	// blockTime, err := gogotypes.TimestampProto(ctx.BlockTime())
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "block time conversion")
+	}
+
+	policy := account.GetDecisionPolicy()
+	if policy == nil {
+		return nil, sdkerrors.Wrap(errors.ErrEmpty, "nil policy")
+	}
+
+	// Prevent proposal that can not succeed.
+	err = policy.Validate(g)
+	if err != nil {
+		return nil, err
+	}
+
+	// Define proposal timout.
+	// The voting window begins as soon as the proposal is submitted.
+	timeout := policy.GetTimeout()
+	window := timeout
+	// endTime, err := gogotypes.TimestampProto(ctx.BlockTime().Add(window))
```

Comment:
why is this commented?

Reply:
we are directly using `ctx.BlockTime().Add(window)` as value for `Timeout` in `Proposal`
So we need not convert it into `gogotypes.Timestamp` as we have decided to update `Timeout` field to not use gogotypes anymore.

Reply:
ok but we still need to set the timeout anyway

Reply:
nevermind just saw it's set directly in `group.Proposal` just below!